### ISSUE TITLE
ot-math: Provide some way to get all defined MathKern correction heights

### ISSUE THREAD

**Issue by lexi-lambda** [harfbuzz/harfbuzz#3396]
The [`hb_ot_math_get_glyph_kerning`](https://harfbuzz.github.io/harfbuzz-hb-ot-math.html#hb-ot-math-get-glyph-kerning) function allows querying a font for a math kern value at a particular correction height. However, it does not provide any way to determine which correction heights at actually defined, and therefore provides no way to efficiently determine where the discontinuities are.

I am working on an OpenType math renderer, and I would like to be able to get access to this information so that my renderer can provide a debugging visualization of the defined cut-in regions for a particular glyph. As far as I can tell, the only way to do this using the current API is to evaluate some very large number of correction heights by calling `hb_ot_math_get_glyph_kerning` once per pixel along the rendered glyph’s y axis, which is obviously not particularly efficient.

Ideally, an API to request this information would look something like [`hb_ot_math_get_glyph_variants`](https://harfbuzz.github.io/harfbuzz-hb-ot-math.html#hb-ot-math-get-glyph-variants) and would allow extracting the entire set of correction heights and corresponding kerning values at once. The interface would look something like this:

```c
unsigned int
hb_ot_math_get_glyph_kernings (hb_font_t *font,
                               hb_codepoint_t glyph,
                               hb_ot_math_kern_t kern,
                               unsigned int *kern_entry_count,
                               hb_ot_math_kern_entry_t *kern_entries);

typedef struct {
  hb_position_t correction_height;
  hb_position_t kern_value;
} hb_ot_math_kern_entry_t;
```

I would be happy to put together a PR to add such a function if this proposal sounds reasonable.

**Comment by behdad** [comment]
Sgtm. Thank you in advance. :)

**Comment by khaledhosny** [comment]
Would it be simpler to have to `hb_position_t` arrays for output instead of the new struct? I don't think we have any API that takes caller-allocated structs for output.

**Comment by behdad** [comment]
> I don't think we have any API that takes caller-allocated structs for output.

`hb_ot_var_get_axis_infos`

**Comment by behdad** [comment]
> ```c
> unsigned int
> hb_ot_math_get_glyph_kernings (hb_font_t *font,
>                                hb_codepoint_t glyph,
>                                hb_ot_math_kern_t kern,
Add `unsigned int start_offset` at this position like other API that fill an array.
>                                unsigned int *kern_entry_count,
>                                hb_ot_math_kern_entry_t *kern_entries);
> ```

**Comment by lexi-lambda** [comment]
I took a stab at implementing this, and I realized that my proposed API can’t be quite right, because there is one more `kern_value` than there are `correction_height`s. That’s because the `correction_height` values morally sit “in between” the `kern_value`s, since they describe the points where the kerning *changes*, there is one “extra” `kern_value` at either the top or the bottom of the glyph (depending on your perspective).

A simple solution to this would be to just add an additional (output) argument to get the extra value, like this:

```c
unsigned int
hb_ot_math_get_glyph_kernings (hb_font_t *font,
                               hb_codepoint_t glyph,
                               hb_ot_math_kern_t kern,
                               hb_position_t *first_kern_value,
                               unsigned int *kern_entry_count,
                               hb_ot_math_kern_entry_t *kern_entries);
```

But I think that’s fairly unhelpful, seeing as it’s no more convenient the results than it is to just consume two separate arrays. So I think I’m leaning towards @khaledhosny’s suggestion to just expose two `hb_position_t` arrays, instead. Incorporating @behdad’s suggestion to include a `start_offset` parameter, that would look like this:

```c
unsigned int
hb_ot_math_get_glyph_kernings (hb_font_t *font,
                               hb_codepoint_t glyph,
                               hb_ot_math_kern_t kern,
                               unsigned int start_offset,
                               unsigned int *kern_value_count,
                               hb_position_t *kern_values,
                               hb_position_t *correction_heights);
```

After some experimentation, I’ve found this representation to be more useful for my needs, anyway (mostly due to the fact that each correction height is really associated with *two* kern values, so tupling them in a struct doesn’t provide much benefit). I’ll probably open a PR implementing this approach soon.
